File: python/base/metadata.py
from typing import Any, Dict, List, Optional

model_to_table: Dict[str, str] = {}
model_fields: Dict[str, Any] = {}
relation_map: Dict[str, Dict[str, Dict[str, str]]] = {}

def set_adapter_metadata(metadata: Dict[str, Any]) -> None:
    model_to_table.clear()
    model_to_table.update(metadata.get("model_to_table") or metadata.get("modelToTable") or {})
    model_fields.clear()
    model_fields.update(metadata.get("model_fields") or metadata.get("modelFields") or {})
    relation_map.clear()
    rels = metadata.get("relation_map") or metadata.get("relationMap") or metadata.get("RELATION_MAP") or {}
    if isinstance(rels, dict):
        for model, relations in rels.items():
            if not isinstance(relations, dict):
                continue
            relation_map[model] = {}
            for key, defn in relations.items():
                if isinstance(defn, dict):
                    relation_map[model][key] = {
                        "modelName": defn.get("modelName") or defn.get("model_name") or "",
                        "relationType": defn.get("relationType") or defn.get("relation_type") or "many",
                        "foreignKey": defn.get("foreignKey") or defn.get("foreign_key") or "",
                        "localKey": defn.get("localKey") or defn.get("local_key") or "",
                    }

def get_relations_for_model(model_name: str) -> Dict[str, Dict[str, str]]:
    return relation_map.get(model_name) or {}

def get_model_to_table() -> Dict[str, str]:
    return dict(model_to_table)

def get_fields_for_model(model_name: str) -> Any:
    return model_fields.get(model_name)

File: golang/base/metadata.go
// Package base provides core dialect types, quoting utilities, vector math,
// and adapter metadata for AN5 Go Adapters.
package base

import (
	"strings"
	"sync"
)

// RelationDef describes a model relation for eager-loading (include) support.
type RelationDef struct {
	ModelName    string
	RelationType string // "many" | "one"
	ForeignKey   string
	LocalKey     string
}

// AdapterMetadata holds model-to-table, field mapping, and relation mapping injected by generated clients.
type AdapterMetadata struct {
	ModelToTable map[string]string
	ModelFields  map[string]interface{}
	RelationMap  map[string]map[string]RelationDef
}

var (
	metaMu       sync.RWMutex
	modelToTable = map[string]string{}
	modelFields  = map[string]interface{}{}
	relationMap  = map[string]map[string]RelationDef{}
)

// SetAdapterMetadata stores metadata provided by the generated an5Client.
// Call this once at application startup after initializing the DB context.
func SetAdapterMetadata(meta AdapterMetadata) {
	metaMu.Lock()
	defer metaMu.Unlock()
	modelToTable = make(map[string]string, len(meta.ModelToTable))
	for k, v := range meta.ModelToTable {
		modelToTable[k] = v
	}
	modelFields = make(map[string]interface{}, len(meta.ModelFields))
	for k, v := range meta.ModelFields {
		modelFields[k] = v
	}
	relationMap = make(map[string]map[string]RelationDef, len(meta.RelationMap))
	for model, rels := range meta.RelationMap {
		relationMap[model] = make(map[string]RelationDef, len(rels))
		for key, def := range rels {
			relationMap[model][key] = def
		}
	}
}

// GetModelToTable returns the model-to-table mapping (read-only snapshot).
func GetModelToTable() map[string]string {
	metaMu.RLock()
	defer metaMu.RUnlock()
	out := make(map[string]string, len(modelToTable))
	for k, v := range modelToTable {
		out[k] = v
	}
	return out
}

// GetFieldsForModel returns field metadata for the given model name.
func GetFieldsForModel(modelName string) interface{} {
	metaMu.RLock()
	defer metaMu.RUnlock()
	return modelFields[modelName]
}

// GetRelationsForModel returns the relation map for the given model name (read-only snapshot).
func GetRelationsForModel(modelName string) map[string]RelationDef {
	metaMu.RLock()
	defer metaMu.RUnlock()
	rels := relationMap[modelName]
	if rels == nil {
		return map[string]RelationDef{}
	}
	out := make(map[string]RelationDef, len(rels))
	for k, v := range rels {
		out[k] = v
	}
	return out
}

// GetRelationMap returns the full model-to-relations map (read-only snapshot).
func GetRelationMap() map[string]map[string]RelationDef {
	metaMu.RLock()
	defer metaMu.RUnlock()
	out := make(map[string]map[string]RelationDef, len(relationMap))
	for model, rels := range relationMap {
		out[model] = make(map[string]RelationDef, len(rels))
		for k, v := range rels {
			out[model][k] = v
		}
	}
	return out
}

// ResolveTable looks up the physical table name for a model name (case-insensitive fallback).
func ResolveTable(modelName string) string {
	metaMu.RLock()
	defer metaMu.RUnlock()
	if t, ok := modelToTable[modelName]; ok {
		return t
	}
	// lowercase fallback
	lower := strings.ToLower(modelName)
	for k, v := range modelToTable {
		if strings.ToLower(k) == lower {
			return v
		}
	}
	return modelName
}
